### PR TITLE
Add keypair.

### LINKED ISSUE #2714: Use KeyPair instead of private/public key
In some signatures algorithms the public key isn't derivable from the Private key. To have more flexibility instead of having always a PrivateKey and derive the PublicKey from it we should have a KeyPair structure so that we can always have access to both

### PR DESCRIPTION
**Pull request by AurelienFT**
Fix #2714

### CODE REVIEW

File: massa-signature/src/signature_impl.rs [hunk 1]
```diff
@@ -20,6 +20,22 @@ const PRIVATE_KEY_STRING_PREFIX: &str = "PRI";
...
+pub struct KeyPair {
+    /// Private key used for signing, wallet dump/load etc...
+    pub private_key: PrivateKey,
+    /// Public key used for verifications
+    pub public_key: PublicKey
+}
+
+impl KeyPair {
+    /// Create a KeyPair from a private key
+    pub fn from_private_key(private_key: PrivateKey) -> KeyPair {
```
Comment (damip) [inline massa-signature/src/signature_impl.rs:34]: this should be removed. Everywhere we were using/storing a private key only, we should be using a keypair

File: massa-signature/src/signature_impl.rs [hunk 2]
```diff
@@ -20,6 +20,22 @@ const PRIVATE_KEY_STRING_PREFIX: &str = "PRI";
 const PUBLIC_KEY_STRING_PREFIX: &str = "PUB";
 const SIGNATURE_STRING_PREFIX: &str = "SIG";
 
+/// `KeyPair` is used for signature and decrypting
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct KeyPair {
```
Comment (damip) [inline massa-signature/src/signature_impl.rs:25]: I think the shnorr crate has a KeyPair ! 
see the definition of our current PrivateKey, it already wraps a keypair

File: massa-signature/src/signature_impl.rs [hunk 3]
```diff
@@ -899,13 +902,17 @@ pub fn verify_signature(
 }
 
 /// Generate a random private key from a RNG.
-pub fn generate_random_private_key() -> PrivateKey {
+pub fn generate_random_keypair() -> KeyPair {
     use secp256k1::rand::rngs::OsRng;
     let mut rng = OsRng::new().expect("OsRng");
-    PrivateKey(secp256k1::KeyPair::from_secret_key(
+    let keypair = secp256k1::KeyPair::from_secret_key(
```
Comment (damip) [inline massa-signature/src/signature_impl.rs:908]: the lib has a native keypair generator: https://docs.rs/secp256k1/latest/secp256k1/struct.KeyPair.html#examples

Comment (damip) [reply massa-signature/src/signature_impl.rs:908]: generally, we shouldn't be using key derivation anymore

File: massa-signature/src/signature_impl.rs (hunk 2 above)
Comment (damip) [reply massa-signature/src/signature_impl.rs:25]: It shouldn't add any constraints: the wrapped type should be private and only a couple methods should be exposed.

For example:

* struct PublicKey(lib::PubKey)
  * verify(data, signature) -> bool

* struct KeyPair(lib::KeyPair)
  * sign(data) -> Signature
  * get_public_key() -> &PublicKey

File: massa-signature/src/signature_impl.rs (hunk 3 above)
Comment (AurelienFT) [reply massa-signature/src/signature_impl.rs]: I changed to use it.

File: massa-signature/src/signature_impl.rs (hunk 1 above)
Comment (AurelienFT) [reply massa-signature/src/signature_impl.rs]: Done.

File: massa-signature/src/signature_impl.rs (hunk 2 above)
Comment (AurelienFT) [reply massa-signature/src/signature_impl.rs]: I have implemented this interface.

**Comment by AurelienFT** [comment]
@damip I have made a new implementation closer to what you asked. Tell me if it's ok so that I can propagate to the rest of the code.

**Comment by damip** [comment]
@AurelienFT Yes it's a welcome improvement !

**Review by Eitu33 — DISMISSED** [review]

**Comment by Eitu33** [comment]
Nice

**Comment by damip** [comment]
don't forget to dump both the pubkey and the privkey in files as well (eg. wallet, node privkey, staking keys etc...)

**Review by Eitu33 — APPROVED** [review]

**Review by Eitu33 — APPROVED** [review]

**Comment by AurelienFT** [comment]
bors merge

**Comment by AurelienFT** [comment]
bors cancel

**Review by Eitu33 — APPROVED** [review]
Signature update LGTM 👍

File: massa-wallet/src/lib.rs [hunk 4]
```diff
@@ -36,7 +36,7 @@ impl Wallet {
     pub fn new(path: PathBuf, password: String) -> Result<Wallet, WalletError> {
         if path.is_file() {
             let content = &std::fs::read(&path)?[..];
-            let decrypted_content = decrypt(&password, content)?;
+            let (__version, decrypted_content) = decrypt(&password, content)?;
```
Comment (Eitu33) [inline massa-wallet/src/lib.rs:39]:
```suggestion
            let (_version, decrypted_content) = decrypt(&password, content)?;
```

**Review by Eitu33 — APPROVED** [review]

**Review by damip — APPROVED** [review]

**Comment by damip** [comment]
bors merge

**Pull request merged**